Fix URL scheme parsing and TLS Server Name Indication (SNI) and its use in `apps/`

DDvO opened the pull request:
~Add recommendation that TLS clients should use `SSL_set_tlsext_host_name()` for SNI
and that this should be done jointly with using `{SSL,X509_VERIFY_PARAM}_{set1,add1}_host()`.
On this occasion fix nits and omissions in the respective `.pod` files.~ ~Doc changes moved to #29150.~

* ~Fix `OSSL_parse_url()` going too far when scanning for '`@`' to detect the end of the `userinfo` component.,
going astray when an `@` symbol occurs after the `host` component, for instance in the `query` component.~ meanwhile superseded by #30319

* Fix `OSSL_parse_url()` to correctly parse scheme (where only certain chars are allowed)

* Make sure that `app_http_tls_cb()` calls `SSL_set_tlsext_host_name()` for setting the SNI only if a server name (rather than an IP address) is given. 

* Also fix the server name being set in the SNI: it must always be the server actually connecting to, 
while for the CMP CLI app so far this is wrongly tied to the `-tls_host` option.

* Add to `OSSL_HTTP_REQ_CTX_nbio` support for partial content-type string matching

* Extend related doc, e.g., of `SSL_set_tlsext_host_name()` and `OSSL_HTTP_REQ_CTX_set_expected()`.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

DDvO committed:
OSSL_HTTP_REQ_CTX.pod: tweak description of 'expect_asn1' parameter of OSSL_HTTP_REQ_CTX_set_expected()

DDvO committed:
SSL_CTX_set_tlsext_servername_callback.pod: improve doc of SSL_set_tlsext_host_name(); update format

Add recommendation to use it for TLS clients, together with X509_VERIFY_PARAM_{set1,add1}_host()

DDvO committed:
apps/cmp.c: fix use of SSL_set_tlsext_host_name() for SNI and X509_VERIFY_PARAM_set1_host()

DDvO commented:
BTW, since doing the SNI and hostname/address checks is important for correct and secure TLS connections,
how about calling on client side `SSL_set_tlsext_host_name()` and `X509_VERIFY_PARAM_set1_host()` by default?

InfoHunter commented on doc/man3/X509_VERIFY_PARAM_set_flags.pod:
```diff
@@ -145,12 +145,16 @@ To obtain all names start with B<n> = 0 and increment B<n> as long as no NULL
-B<name> clearing any previously specified hostname.  If
-B<name> is NULL, or empty the list of hostnames is cleared, and
+B<name>, clearing any previously specified hostname.
+If B<name> is NULL or empty, the list of hostnames is cleared, and
```
Isn't NULL here equal to empty?

DDvO replied:
At this point in the text I just moved a comma and re-formatted the lines, without semantic changes.
Since `name` is a string pointer, in this context "empty" means a string of zero length (`""`),
which is different from the NULL pointer, but for this function has the same effect.

DDvO replied:
To make crystal clear what is meant by 'empty', I've meanwhile extended it to 'the empty string', 
as was already done in `SSL_set1_host.pod`.

BTW, I just had a look at the implementation of `int_x509_param_set_hosts()`,
which does some strange extra checks on the string input including its (optional) length parameter.

mattcaswell commented on doc/man3/SSL_CTX_set_tlsext_servername_callback.pod:
```diff
@@ -123,22 +123,29 @@ client is processed. The servername, certificate and ALPN callbacks occur after
+
+Using this function may be crucial also for correct routing of connection requests.
+TLS clients are recommended to use this function
+as well as L<X509_VERIFY_PARAM_set1_host(3)> or L<X509_VERIFY_PARAM_add1_host(3)>
```
I would instead recommend the use of `SSL_set1_host` which is more straightforward to use (you don't have to first get hold of the `X509_VERIFY_PARAM` and handles IP addresses more correctly)

DDvO replied:
Good point - done.
So far I was not aware of `SSL_{set1,add1}_host()`.

mattcaswell commented on doc/man3/X509_VERIFY_PARAM_set_flags.pod:
```diff
@@ -145,12 +145,16 @@ To obtain all names start with B<n> = 0 and increment B<n> as long as no NULL
 is NUL-terminated, B<namelen> may be zero, otherwise B<namelen>
 must be set to the length of B<name>.
 
+TLS clients are recommended to use this function or X509_VERIFY_PARAM_add1_host()
```
As above. I would not recommend this function but instead `SSL_set1_host()`

DDvO replied:
Changed also this accordingly.

mattcaswell commented on apps/cmp.c:
```diff
@@ -1510,8 +1517,7 @@ static SSL_CTX *setup_ssl_ctx(OSSL_CMP_CTX *ctx, const char *host,
          */
-        if (!truststore_set_host_etc(trust_store,
-                                     opt_tls_host != NULL ? opt_tls_host : host))
+        if (!truststore_set_host_etc(trust_store, host))
```
Not sure I understand the purpose of this change. Can you explain it?

DDvO replied:
`setup_ssl_ctx()` is called by `setup_client_ctx()` if TLS is used.
The change moves the conditional expression to the latter function
because the hostaddr is now needed also for setting up `info->sni_hostname` there.
This is for correct and consistent use of the host for hostname validation and SNI.

esyr replied:
Is it used for SNI, however? `SSL_set_tlsext_host_name` is called in `app_http_tls_cb`, and it uses `info->server`, which is set to `host` (and not `opt_tls_host`) in `setup_client_ctx`, if I understand correctly.

esyr replied:
Or is it meant the other way, that since it is no longer picked up from the verify params, that SNI is expected to be set to `host`, and `tls_host` is used for cert validation only and not for SNI?

DDvO replied:
`truststore_set_host_etc(trust_store, host)` is meanwhile for the hostname check only, which may be adapted by the `-tls_host` option, as per the updated doc.
Whereas `info->server` holds the hostname used for contacting the server, which is used also for SNI (unless it is an IP address).

t8m commented:
This needs a rebase due to conflicts.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

mattcaswell commented on apps/include/apps.h:
```diff
@@ -45,6 +45,8 @@
  */
 # define _UC(c) ((unsigned char)(c))
 
+# define IS_IP_ADDR(host) ((host) != NULL && ((*(host) >= '0' && *(host) <= '9') || *(host) == '['))
```
Hmm. So according to this `7-zip.org` is an IP address!

vdukhovni replied:
That's deeply wrong. To determine whether a string is an IP address, one needs to pass  to `getaddrinfo(3)` with AI_NUMERICHOST among the hints flags and `AF_UNSPEC` as the address family. This will cover both v4 and v6 addressed. Names enclosed in `[]` are presumptively address literals since OpenSSL does not support these around hostnames (in MTAs these often disable MX lookups, but that's not applicable here).

DDvO replied:
Fixed, introducing instead a proper `APP_is_IP_address()` function along the lines given by Viktor.

vdukhovni reviewed (changes requested):
Please fix at least the "is IP address" test.

DDvO commented:
After experimenting with the outcome of the new check_backports CI,
I removed 3.0 from the target list because unfortunately there is a trivial merge conflict that I cannot work around in this PR.
So a backport to 3.0, if still desirable, will require a separate PR.

I also tried with 3.2, but then noticed that there is some other such merge conflict there,
and since 3.2 will anyway expire soon, I removed the 'branch: 3.2' label - so the above CI failure for 3.2 does no more apply.

DDvO commented:
Thanks @mattcaswell and @vdukhovni for your comments of mid-June on my simplistic `IS_IP_ADDR()` -
I finally took the time to fix this after getting reminded of this a few days back.

DDvO commented:
@vdukhovni did you see that I meanwhile did the suggested fix?

vdukhovni commented on apps/cmp.c:
```diff
@@ -528,7 +528,7 @@ const OPTIONS cmp_options[] = {
     {OPT_MORE_STR, 0, 0, "this implies hostname validation"},
     {"tls_host", OPT_TLS_HOST, 's',
-     "Address to be checked (rather than -server) during TLS hostname validation"},
+     "Address to be used for SNI and to be checked during TLS hostname validation"},
```
```suggestion
     "Name (or address) to use in SNI and verify in the certificate"},
```

vdukhovni commented on crypto/http/http_lib.c:
```diff
@@ -92,8 +92,8 @@ int OSSL_parse_url(const char *url, char **pscheme, char **puser, char **phost,
     user = user_end = host = p;
-    host = strchr(p, '@');
-    if (host != NULL)
+    host = strpbrk(p, "@[]/?#"); /* [ and ] delimit IPv6 addresses */
+    if (host != NULL && *host == '@')
```
I should not that the parsing of `<scheme>://` a few lines above is incorrect.  It misinterprets:
```
host.example.org/some/path#://not-a-scheme/not.a.host:404
```
The scheme name has a [restricted syntax](https://datatracker.ietf.org/doc/html/rfc3986#section-3.1) that SHOULD be checked, at least to not contain ':/@#?', but simpler to check the allowed charset.

vdukhovni commented on doc/man1/openssl-cmp.pod.in:
```diff
@@ -1029,11 +1030,12 @@ The certificate verification options
 have no effect on the certificate verification enabled via this option.
 
-=item B<-tls_host> I<name>
+=item B<-tls_host> I<address>
```
This is a mistake IMHO.  It'll rarely be an IP address, more frequently one might put an address in the URI to deal with NAT or similar, but specify the name in `-tls_host`.  Certificates with IP addresses are quite rare. So this should say `host`, and explain that `host` is typically a DNS name, and sometimes (rarely) an IP address.

vdukhovni commented on doc/man1/openssl-cmp.pod.in:
```diff
@@ -1029,11 +1030,12 @@ The certificate verification options
-Address to be checked during hostname validation.
-This may be a DNS name or an IP address.
-If not given it defaults to the B<-server> address.
+Address to be used for Server Name Indication (SNI) according to RFC 3546 section 3.1
+and to be checked during TLS hostname validation.
+This may be a DNS name (for SNI, only this is used) or an IPv4 or IPv6 address.
+If not given it defaults to the host part of the B<-server> option URL argument.
```
Therefore, this also needs a rewrite, because only DNS names are used with SNI, and addresses are not.
This option is typically only needed when direct connectivity to the desired hostname is not available, or when a specific address of a host is to be tested, in which case the `-server` option URL might specify the address and `-tls_host` would have the SNI name.

Also, in that case, which name ends up copied into HTTP 1.1 "Host:" header?  If the one in the URL, then it is far from clear when `-tls_host` would be useful, since SNI hostname really should match the "Host:" header.

So I am not sure what this is really for.

DDvO replied:
I figured that it would not fit with this bugfix PR to add the `-connect` option (and a related HTTP client API function) here - I'm going to to this in a follow-up PR. So just fixed the existing `-tls_host` option and its documentation to make clear what it does (in comparison with the `host:[port]` given with the `-server` option.

vdukhovni replied:
I am not yet convinced this is best, but I'll think about it. As-is the option does not really make much sense. But if the PR is better than status quo, maybe.  Though it is a rather large bug-fix only PR.  Can something much smaller be the bug fix?

DDvO replied:
> As-is the option does not really make much sense. But if the PR is better than status quo, maybe.

As I wrote above, the `-tls_host` CLI option should stay for backward compatibility.
It can be used for the purpose now better described in the docs,
though I agree that the `-connect` option is going to be more useful.

> Though it is a rather large bug-fix only PR. Can something much smaller be the bug fix?

As its title and OP still says, this PR originally was mostly on documentation improvements, 
plus an SNI-related bug fix (namely, that the host info must be used for SNI only in case of DNS names),
plus a fix of URI parsing.
@vdukhovni due to your review comments, the URI parsing received further fixes, and the IP address (vs. domain name) recognition got improved, which added a new app-internal function: `APP_is_IP_address()`.
I see little opportunity for downsizing the code changes needed for fixing the mentioned bugs.

DDvO committed:
http_lib.c: fix parsing 'scheme' part in OSSL_parse_url()

DDvO commented:
Ping @openssl/committers for further reviews/approvals, please.

DDvO commented:
Ping again @vdukhovni

t8m commented on doc/man3/X509_VERIFY_PARAM_set_flags.pod:
```diff
@@ -311,6 +315,8 @@ requirements and lead to a locally trusted root.
 explicit trust settings (see "TRUST SETTINGS" in L<openssl-x509(1)>).
 
+=begin comment
+
```
Do not just comment out but remove. However this sounds like something for a separate PR targeting just master branch.

DDvO commented on apps/lib/apps.c:
```diff
@@ -2552,6 +2555,31 @@ void store_setup_crl_download(X509_STORE *st)
+    hints.ai_flags = AI_NUMERICHOST;
+    ret = getaddrinfo(host, NULL, &hints, &res);
+    if (res != NULL)
+        freeaddrinfo(res);
```
I had got the hint to use `getaddrinfo()` from https://github.com/openssl/openssl/pull/27357#discussion_r2150153588.

I did not find how to use here `BIO_ADDR` (or `BIO_lookup_ex()`, which is so far the only OpenSSL function calling `getaddrinfo()`), but found another simple solution using `a2i_IPADDRESS()`.
If this is not sufficient, please provide a better (more specific) hint.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

DDvO committed:
http_lib.c: add check that host_end is not past authority_end

DDvO commented on crypto/http/http_lib.c:
Good point to add such a check on this occasion - added a commit for it.

DDvO commented on crypto/http/http_lib.c:
```diff
@@ -109,6 +114,8 @@ int OSSL_parse_url(const char *url, char **pscheme, char **puser, char **phost,
             goto parse_err;
         p = ++host_end;
+        if (host_end > authority_end)
+            goto parse_err;
```
Nice simplification - done.

vdukhovni commented on crypto/http/http_lib.c:
```diff
@@ -105,7 +110,7 @@ int OSSL_parse_url(const char *url, char **pscheme, char **puser, char **phost,
     if (host[0] == '[') {
         /* IPv6 literal, which may include ':' */
-        host_end = strchr(host + 1, ']');
+        host_end = memchr(host + 1, ']', authority_end - host);
```
FWIW, with `host + 1`, you're technically off by one in the length, and the `memchr()` may also (harmlessly) peek at a terminating `NUL` byte to compare it with `]` (if there is no path component in the URL, otherwise it might look tag the path separator, one of: `/#?`).
The pedantically correct length is 1 byte less. That's why I originally suggested to just to start at `host` even if that redundantly looks at the initial `[` again, otherwise (with `host + 1`) you could use `authority_end - host - 1` for the length.
The suggestion below is optional (keeps the code simpler I think), but you have now is still correct, despite the 1-byte extra read in broken URLs with no terminating `]`.  Do what you think is best.
```suggestion
        host_end = memchr(host, ']', authority_end - host);
```

DDvO replied:
Oh, right. 
I went for the slightly more efficient and fully correct version, though it requires the extra `- 1` in the length calculation.

DDvO commented:
Rebased again to fix (simple) doc merge conflict

DDvO commented:
Rebased again, fixing (trivial) new doc merge conflict.

vdukhovni commented on apps/cmp.c:
```diff
@@ -472,13 +472,13 @@ const OPTIONS cmp_options[] = {
+        "[http[s]://]host[:port][/path] of CMP server to use. Default port 80 or 443." },
     { OPT_MORE_STR, 0, 0,
-        "address may be a DNS name or an IP address; path can be overridden by -path" },
+        "host may be DNS name or IP address; path can be overridden by -path" },
```
```suggestion
        "host may be a DNS name or an IP address; path can be overridden by -path" },
```

DDvO replied:
Done, as well as the following two text improvements you indicated below.

vdukhovni commented on apps/cmp.c:
```diff
@@ -861,7 +861,11 @@ static X509 *load_cert_pwd(const char *uri, const char *source, const char *desc
+/*
+ * Set expected hostname/IP address and clears any email address in the given ts.
+ * If the host is NULL, host name/address verification is disabled.
+ * Otherwise, it is interpreted as an IP address if possible, otherwise as a domain name.
```
```suggestion
 * It is interpreted as an IP address when possible, otherwise as a domain name.
```

vdukhovni commented on apps/cmp.c:
```diff
@@ -871,9 +875,12 @@ static int truststore_set_host_etc(X509_STORE *ts, const char *host)
         X509_CHECK_FLAG_ALWAYS_CHECK_SUBJECT | X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS);
-    return (host != NULL && X509_VERIFY_PARAM_set1_ip_asc(ts_vpm, host))
+    return X509_VERIFY_PARAM_set1_ip_asc(ts_vpm, host)
         || X509_VERIFY_PARAM_set1_host(ts_vpm, host, 0);
```
@bob-beck Is this an adequate idiom, or should the caller first parse the address, and then call the right half of the `||`?

DDvO replied:
Following the below hint given by Claude AI, improved this to:
```
return host_is_ip_address(host)
        ? X509_VERIFY_PARAM_set1_ip_asc(ts_vpm, host)
        : X509_VERIFY_PARAM_set1_host(ts_vpm, host, 0);
```

vdukhovni commented on doc/man1/openssl-cmp.pod.in:
```diff
@@ -527,9 +527,14 @@ Reason numbers defined in RFC 5280 are:
+If the I<host> string is an IPv6 address, it must be enclosed in C<[> and C<]>.
+
+For TLS connections the I<host> is verified in TLS server certificate
+unless the B<-tls_host> option is given.
```
```suggestion
For TLS connections, the name verified in server certificates is I<host>
unless the B<-tls_host> option is used to specify a different name.
```

vdukhovni reviewed (changes requested):
I'm becoming concerned that the "and its use in apps" part is making this this PR a bit difficult to review. It would have been easier as multiple PRs.

Perhaps with AI help I'll be able to digest the whole thing, but let's try to avoid big chunks like this going forward.

vdukhovni commented:
I asked Claude to double-check my comments.

## feedback on my comments

* The two wording fixes are straightforward. "when possible" flows better than "if possible" in that comment.

* The pod rewrite is better than simply inserting the missing "the" — "the name verified in server certificates is `host` unless `-tls_host` is used to specify a different name" says what actually happens rather than describing it passively.

* The `||` question to Bob is the most substantive. `X509_VERIFY_PARAM_set1_ip_asc` returns 0 for two distinct reasons: the string is not a valid IP address, and a memory allocation failure. The `||` idiom conflates them — if the string is a valid IP but allocation fails, the fallthrough calls `set1_host` with an IP string as a hostname, silently swallowing the error. In practice this is harmless (no valid certificate carries an IP as a DNS SAN, so verification would correctly fail), but it takes the wrong path for the wrong reason.

Pre-parsing first with `host_is_ip_address()` — which this PR already introduces for the SNI decision — would make the branching explicit and propagate genuine errors correctly. That said, the `||` pattern predates this PR, so Bob's read on whether it is considered adequate convention is worth having before pushing for a change here.

## Claude's review of the PR as a whole.

Overall the changes are correct; a few items worth addressing.

---

**`f35afcd` — scheme parsing**

The fix is correct and well-tested. `strstr(url, "://")` matching anywhere in the string (including inside a fragment) is the real bug, and the RFC 3986 §3.1 walk fixes it cleanly.

Minor naming nit: `OSSL_SCHEME_POSTFIX` — the standard term for the trailing part of a fixed string is *suffix*, not *postfix*. Suggest renaming to `OSSL_URL_SCHEME_SUFFIX`. On scope: a slight preference for defining it at file scope after the includes, or, if kept local, adding `#undef OSSL_URL_SCHEME_SUFFIX` before leaving the block — but not a requirement either way.

---

**`f16369f` — IPv6 bounds check**

Correct. `memchr(..., authority_end - host - 1)` constrains the `]` search to the authority; the old `strchr` would accept a `]` in the path. The length argument is safe: `host[0] == '['` implies `authority_end > host`.

---

**`c4b4b98` — SNI and hostname verification in `apps/cmp.c`**

Correct fix on both counts. Retrieving the hostname back from `X509_VERIFY_PARAM_get0_host()` to pass to `SSL_set_tlsext_host_name()` was wrong: that getter returns only a DNS-name entry (IP verification uses a separate param), and SNI must not be sent for IP address destinations (RFC 6066 §3). Using `info->server` directly and guarding with `host_is_ip_address()` is the right approach. The `opt_tls_host` override is now correctly applied for cert verification while SNI uses the actual server name. Moving `OPENSSL_free(info->server/port)` into `APP_HTTP_TLS_INFO_free()` is a good consolidation.

---

**`7ebe8b6` — content-type / ASN.1 documentation**

The code change is cosmetic only (reformatting a long `OPENSSL_strncasecmp` call onto one line; no logic change). The partial content-type matching was already implemented. The commit title "add support for partial content-type string matching" does not reflect this — something like "OSSL_HTTP_transfer.pod: document expected_content_type and expect_asn1 parameters" would be accurate.

The documentation added to `OSSL_HTTP_transfer.pod` is a near-verbatim copy of text already in `OSSL_HTTP_REQ_CTX.pod`. The original cross-reference a few lines below — "For the meaning of the `expected_content_type`, `expect_asn1`, `timeout`..." — is now redundant and slightly contradictory. Either remove the cross-reference, or replace the duplicated text with a pointer to `OSSL_HTTP_REQ_CTX_new(3)`.

---

**`78e255b` and `c4b4b98` — documentation phrasing**

In `openssl-cmp.pod.in`: "verified in TLS server certificate" should be "verified in **the** TLS server certificate".

In `SSL_CTX_set_tlsext_servername_callback.pod`, the new paragraph reads:

> Using this function may be crucial also for correct routing of connection requests.
> TLS clients are recommended to use this function as well as ...

Suggested wording:

> Using this function may also be important for correct routing of connection requests.
> TLS clients should call this function alongside `SSL_set1_dnsname(3)` or
> `SSL_add1_dnsname(3)` to set the expected server hostname(s) for certificate validation.

The `B<cb>` → `I<cb>`, `B<al>` → `I<al>`, `B<arg>` → `I<arg>` corrections elsewhere in the same file are correct pod convention.

DDvO commented:
Thanks @vdukhovni for your further comments.
Good thought to use also Claude for reviewing this PR (besides your comments), 
which confirmed the soundness of this PR while bringing up further tweaks.
I've just handled all of them.

t8m commented:
@DDvO Can you please create the OSSL_parse_url() change backport PR for 4.0 and below?

I am fine with the changes but I am hesitant backporting all of them to stable branches.

DDvO commented:
> @DDvO Can you please create the OSSL_parse_url() change backport PR for 4.0 and below?
> 
> I am fine with the changes but I am hesitant backporting all of them to stable branches.

No problem, done here: https://github.com/openssl/openssl/pull/31267

DDvO commented:
When aligning this PR with a related one in a different project, 
I got an interesting Copilot review comment pointing out a problem with the new `host_is_ip_address()` helper function:
it does not cope with IPv6 addresses enclosed in `[` and `]`.

In retrospect, it would have been better if `OSSL_*parse_url()` returned the `host` component without these surrinding chars, which are needed only at URL level.
Yet this has been implemented and documented otherwise, so little chance to rectify this now.

So I had to push an extension of `host_is_ip_address()` here that glosses over those brackets if present.
and ask @t8m to reconfirm approval after this extra fix.

esyr commented on doc/man3/SSL_CTX_set_tlsext_servername_callback.pod:
```diff
@@ -26,9 +26,9 @@ the ClientHello callback, which can be set using SSL_CTX_set_client_hello_cb().
+SSL_CTX_set_tlsext_servername_callback() sets the application callback I<cb>
 used by a server to perform any actions or configuration required based on
-the servername extension received in the incoming connection. When B<cb>
+the servername extension received in the incoming connection. When I<cb>
```
It's weird that `servername` is not emphasised via `B<>`/`C<>`, but okay.

DDvO replied:
Added `B<  ... >`

esyr commented on doc/man3/SSL_CTX_set_tlsext_servername_callback.pod:
```diff
@@ -127,22 +127,28 @@ client is processed. The servername, certificate and ALPN callbacks occur after
+SSL_set_tlsext_host_name() sets the Server Name Indication (SNI)
+ClientHello extension defined in RFC 3546 section 3.1
+to contain the value I<name> of type B<TLSEXT_NAMETYPE_host_name>.
+If I<name> is NULL it clears the SNI extension.
```
`s/ it/, it/`

DDvO replied:
Fixed, as well as the other doc/comment nits.

esyr commented on doc/man3/SSL_CTX_set_tlsext_servername_callback.pod:
```diff
@@ -127,22 +127,28 @@ client is processed. The servername, certificate and ALPN callbacks occur after
+L<SSL_add1_dnsname(3)> to set the expected server hostname(s) for certificate validation.
+
+The SSL_set_tlsext_host_name() function should only be called on SSL objects
+that will act as clients; otherwise the configured I<name> will be ignored.
```
`s/otherwise/otherwise,/`

esyr commented on crypto/http/http_lib.c:
```diff
@@ -79,16 +80,20 @@ int OSSL_parse_url(const char *url, char **pscheme, char **puser, char **phost,
-        if (scheme_end == scheme)
-            goto parse_err;
-        p += strlen("://");
+    /* check for optional prefix "<scheme>://" as per RFC 3986: */
```
`s/: / /`

esyr commented on apps/cmp.c:
```diff
@@ -871,10 +875,18 @@ static int truststore_set_host_etc(X509_STORE *ts, const char *host)
+    if (host == NULL)
+        return 1;
+
+#ifndef OPENSSL_NO_SOCK
```
Maybe it has been decided on before, but why exactly is it conditioned on `sock` disablable? Host/IP address are just strings, and they are matched against information in certificates, so, networking shouldn't be required (and I don't see implicit name resolution here so far), or am I missing something?

DDvO replied:
This guard was mostly due to the earlier implementation of `host_is_ip_address()`.
With the current one, the only blocking point was that `INET6_ADDRSTRLEN` was not defined with `OPENSSL_NO_SOCK`, which is not defined due to  `#include <netinet/in.h>` not being done.
As you suggested below, I've now added
```
#ifndef INET6_ADDRSTRLEN /* not defined on OPENSSL_NO_SOCK */
#define INET6_ADDRSTRLEN 46
#endif
```
so we can now dispense with the that guard, also for `host_is_ip_address()` in `apps.c`.

esyr commented on apps/lib/apps.c:
```diff
@@ -2868,6 +2870,31 @@ void store_setup_crl_download(X509_STORE *st)
+    str = a2i_IPADDRESS(host);
+    ERR_pop_to_mark();
+    ASN1_OCTET_STRING_free(str);
+    return str != NULL;
```
This is a UB. Even if you don't dereference `str`, using invalid pointers is still UB, as, per paragraph 2 of section 6.2.4 of ISO 9899:1999, "the value of a pointer becomes indeterminate when the object it points to reaches the end of its lifetime", and equality operator is defined only for pointers that point to an object (and some additional cases, such as NULL).

DDvO replied:
All right, introduced an aux variable to be able to check the `str` for non-NULL before freeing it.

esyr commented on apps/lib/apps.c:
```diff
@@ -2868,6 +2870,31 @@ void store_setup_crl_download(X509_STORE *st)
     X509_STORE_set_lookup_crls_cb(st, crls_http_cb);
 }
 
+#ifndef OPENSSL_NO_SOCK
```
Again, not sure why is this guard, the check should be just an operation on strings.

DDvO replied:
See my response above.

esyr commented on doc/man1/openssl-cmp.pod.in:
```diff
@@ -527,9 +527,14 @@ Reason numbers defined in RFC 5280 are:
-IP address may be for v4 or v6, such as C<127.0.0.1> or C<[::1]> for localhost.
-If the host string is an IPv6 address, it must be enclosed in C<[> and C<]>.
+of the CMP server to connect to via HTTP(S).
+An IP address may be for v4 or v6, such as C<127.0.0.1> or C<[::1]> for C<localhost>.
+If the I<host> string is an IPv6 address, it must be enclosed in C<[> and C<]>.
```
Not sure why is it stated here explicitly, as it is part of RFC 3986.

DDvO replied:
Just for clarity. 
As a user, I would not have been aware of that, and likely many others, neither.

esyr commented on crypto/http/http_lib.c:
```diff
@@ -105,7 +110,7 @@ int OSSL_parse_url(const char *url, char **pscheme, char **puser, char **phost,
     if (host[0] == '[') {
         /* IPv6 literal, which may include ':' */
-        host_end = strchr(host + 1, ']');
+        host_end = memchr(host + 1, ']', authority_end - host - 1);
```
I guess there could be a regression test for this change as well.

DDvO replied:
Added one.

esyr commented on apps/lib/apps.c:
```diff
@@ -50,6 +50,8 @@
 static int WIN32_rename(const char *from, const char *to);
 #define rename(from, to) WIN32_rename((from), (to))
+#include <winsock2.h> /* for INET6_ADDRSTRLEN */
+#include <ws2tcpip.h> /* for INET6_ADDRSTRLEN */
```
If that's the only reason for `OPENSSL_NO_SOCK` guards, it can be hard-coded to 46 (or, well, a fallback definition can be provided), I don't think anyone expect it to ever change.

DDvO replied:
Fine, did so.

esyr reviewed (changes requested):
Generally looks good, the only blocker is UB.

esyr commented on test/http_test.c:
```diff
@@ -385,7 +398,8 @@ static int test_http_url_ipv4(void)
 {
-    return test_http_url_ok("http://[FF01::101]:6", 0, "[FF01::101]", "6", "/");
+    return test_http_url_ok("http://[FF01::101]:6", 0, "[FF01::101]", "6", "/")
+        && !test_http_url_ok("http://[FF01::101/path]", 0, "[FF01::101]", "80", "/path]");
```
Hm, why not use `test_http_url_invalid()` here? The way it's written now is somewhat misleading, I've initially overlooked the negation and assumed that it successfully decodes into `"[FF01::101]"` and `"/path]"`.

DDvO replied:
Good point, I had overlooked the presence of `test_http_url_invalid()`. 
Using it now.

esyr reviewed (changes requested):
LGTM now, thanks, the only nit is that the added test should probably rather use `test_http_url_invalid()`.

esyr reviewed (dismissed):
LGTM, thank you!

nhorman commented on apps/lib/apps.c:
```diff
@@ -2935,27 +2960,32 @@ BIO *app_http_tls_cb(BIO *bio, void *arg, int connect, int detail)
     return bio;
+
+err:
+    BIO_free(sbio);
```
Is it safe to free the BIO here if the SSL object has it set above?

DDvO replied:
I'd say yes, it should be safe because 
* the `BIO_set_ssl(sbio, ssl, BIO_CLOSE)` call transfers the ownership of `ssl` to `sbio`, and
* in the `err` branch, `BIO_free(sbio)` frees `ssl` along with `sbio`.

Both of which (as often, unfortunately) is not documented.

nhorman replied:
wait, I had this backwards (I had thought the call to BIO_set_ssl was actuall a call to SSL_set_bio).  But, given that, if BIO_set_ssl fails/returns 0, then don't we leak the ssl we just created?   I grant thats going to be an unlikely event (it can happen if ssl_new or BIO_up_ref fails).

It looks like thats actually a pretty consistent pattern in uses of BIO_set_ssl.

DDvO replied:
@nhorman can you then also approve?

nhorman replied:
sorry, I forgot to submit my last round of reviews.  As noted above, I had the problem backwards.  You're right in that BIO_free will free the SSL, but if BIO_set_ssl fails, we will leak the allocated SSL.  not sure how big a deal that is (it seems to be a fairly common pattern), but it definately looks like a leak problem.

DDvO replied:
Now I see what you mean: what happens if the `BIO_set_ssl()` call fails?

I just had a closer look at `BIO_set_ssl()` and found that there is one case where `ssl_ctrl()` in `ssl/bio_ssl.c` returns 0 *after* transferring ownership (via `bs->ssl = ssl`), namely if `BIO_up_ref(bio))` fails, which is very unlikely but possible.
Still, so far one cannot fully rely on the ownership transfer to happen if and only if `BIO_set_ssl()` succeeds,
so calling it, one either risks a mem leak (if ignoring its negative result) or a double free (if deallocating `ssl` on its failure).

To fully solve this, I now deallocate `ssl` in case `BIO_set_ssl()` fails
while adding a little bugfix commit that 
* makes sure that `BIO_set_ssl()` transfers ownership of `ssl` only on success, and
* documents this

DDvO replied:
> I guess you and Neil have just opened a can of worms: if I'm not mistaken, failing `ssl_new()` (due to failing `OPENSSL_zalloc`) after `ssl_free` would lead to `BIO_SSL` double-free, as `b->ptr` is not reset to `NULL` after `OPENSSL_free` call in `ssl_free`, and the latter would be called the second time in [1].

@esyr which `b->ptr` in which function context do you mean here?
Since I've changed `ssl_ctrl()` to set `bs->ssl` only at the end on success,
and as I call `SSL_free(ssl)` only on failure, I do not see that I introduced potential for double-free.
Which does not preclude that the wonderful code in `bio_lib.c` bears other potential for double-free,
but that's unrelated to this PR.

nhorman replied:
I think what @esyr is getting at is the following set of conditions:
1) BIO_set_ssl is called
2) In ssl_ctrl, we fall into the BIO_C_SET_SSL case
3) ssl_free is called on the passed bio b
4) ssl_free(b) is called
5) ssl_free calls BIO_get_data to get the data bio (bs)
6) SSL_free(bs->ssl) is called
7) OPENSSL_free(bs) is called.  Critically the BIO data pointer remains pointed to bs (from (5)) which is now no longer valid
8) back in ssl_ctrl, ssl_new is called, which we assume fails on a malloc failure, meaning a new bs is not allocated
9) ssl_ctrl returns with an error code and the BIOS bs/data pointer still pointing to invalid memory (due to the free in (7)
10) your code above detects the error, frees the previously leaked SSL (so you've fixed that and jumps to err:
11) at the err label BIO_free is called
12) BIO_free calls ssl_free, which calls BIO_get_data to get bs again, which is non-null but invalid.
13) we try to call OPENSSL_free again on bs, which results in a double free.

If thats the case it seems like the best solution here is to call BIO_set_data(b, NULL) in ssl_free to set the freed pointer to NULL, and then check it at the top of ssl_free to not use  it if it is NULL.

That all said, given the proliferation of the call sites for this, and the fact that no one currently checks the return code of the BIO_set_ssl call anywhere, this might be creeping out of scope for this PR.  Maybe just let it be and open an issue to fix this separately along with all the call sites.

esyr replied:
Yes, that what I was trying to tell.  In general, I concur that having broken `BIO_set_ssl()` is out of scope for this PR. Created [1] to track this.

[1] https://github.com/openssl/openssl/issues/31388

DDvO committed:
fixup! apps/cmp.c: fix use of SSL_set_tlsext_host_name() for SNI and X509_VERIFY_PARAM_set1_host()

DDvO committed:
make sure BIO_set_ssl() transfers ownership of ssl only on success; document this

esyr commented on ssl/bio_ssl.c:
```diff
@@ -299,7 +299,6 @@ static long ssl_ctrl(BIO *b, int cmd, long num, void *ptr)
         }
```
I guess you and Neil have just opened a can of worms:  if I'm not mistaken, failing `ssl_new()` (due to failing `OPENSSL_zalloc`) after `ssl_free` would lead to `BIO_SSL` double-free, as `b->ptr` is not reset to `NULL` after `OPENSSL_free` call in `ssl_free`, and the latter would be called the second time  in [1].

[1] https://github.com/openssl/openssl/pull/27357/files#diff-22af71ddb18c32051f7e626ab7089fb69c056a419dd209f6c0c85a873e30c2f5R2983

esyr commented on ssl/bio_ssl.c:
```diff
@@ -311,6 +310,7 @@ static long ssl_ctrl(BIO *b, int cmd, long num, void *ptr)
             BIO_set_next(b, bio);
         }
         BIO_set_init(b, 1);
+        bs->ssl = ssl;
```
The change looks innocuous enough (famous last words), however it's slowly creeping out of scope of the PR.

@mattcaswell might also want to take a look at this.

openssl-machine commented:
This pull request is ready to merge

DDvO commented:
Merged - thanks @nhorman, @esyr, and @t8m for your approvals 
and @InfoHunter, @mattcaswell, and @vdukhovni for your reviews!
